workbench-core/environments/src/services/environmentService.test.ts

```diff
@@ -1042,5 +1042,104 @@ describe('EnvironmentService', () => {
         error: undefined
       });
     });
+    test('failed because ETC does not exist', async () => {
+      // BUILD
+      // Get env metadata
+      const batchItems = {
+        Responses: {
+          [TABLE_NAME]: [
+            marshall({ ...projItem, resourceType: 'project' }),
+            marshall({ ...datasetItem, resourceType: 'dataset' })
+          ] // Order is important
+        }
+      };
+      // @ts-ignore
+      ddbMock.on(BatchGetItemCommand).resolves(batchItems);
+
+      // Write data to DDB
+      ddbMock.on(TransactWriteItemsCommand).resolves({});
+
+      // Get environment from DDB
+      const metaData = [datasetItem, envTypeConfigItem, projItem];
+      const envWithMetadata = [env, ...metaData];
+      const queryItemResponse: QueryCommandOutput = {
+        Items: envWithMetadata.map((item) => {
+          return marshall(item);
+        }),
+        $metadata: {}
+      };
+      ddbMock.on(QueryCommand).resolves(queryItemResponse);
+
+      // OPERATE && CHECK
+      await expect(envService.createEnvironment(createEnvReq, authenticateUser)).rejects.toThrow(
+        'envTypeId envType-123 with envTypeConfigId envTypeConfig-123 does not exist'
+      );
+    });
+    test('failed because Project does not exist', async () => {
+      // BUILD
+      // Get env metadata
+      const batchItems = {
+        Responses: {
+          [TABLE_NAME]: [
+            marshall({ ...envTypeConfigItem, resourceType: 'envTypeConfig' }),
+            marshall({ ...datasetItem, resourceType: 'dataset' })
+          ] // Order is important
+        }
+      };
+      // @ts-ignore
+      ddbMock.on(BatchGetItemCommand).resolves(batchItems);
+
+      // Write data to DDB
+      ddbMock.on(TransactWriteItemsCommand).resolves({});
+
+      // Get environment from DDB
+      const metaData = [datasetItem, envTypeConfigItem, projItem];
+      const envWithMetadata = [env, ...metaData];
+      const queryItemResponse: QueryCommandOutput = {
+        Items: envWithMetadata.map((item) => {
+          return marshall(item);
+        }),
+        $metadata: {}
+      };
+      ddbMock.on(QueryCommand).resolves(queryItemResponse);
+
+      // OPERATE && CHECK
+      await expect(envService.createEnvironment(createEnvReq, authenticateUser)).rejects.toThrow(
+        'projectId proj-123 does not exist'
+      );
+    });
+    test('failed because Dataset does not exist', async () => {
+      // BUILD
+      // Get env metadata
+      const batchItems = {
+        Responses: {
+          [TABLE_NAME]: [
+            marshall({ ...envTypeConfigItem, resourceType: 'envTypeConfig' }),
```

Comment:
Looks like there might be an opportunity to do a refactor here using a `beforeEach` where you can define `batchItems`. That would prevent the copy-paste because it looks like the setup is the same and `batchItems` is the only difference. Not a blocker though.

Reply:
updated